samza-core/src/main/java/org/apache/samza/task/TaskCallbackManager.java

```diff
@@ -46,22 +49,22 @@
      * Adding the newly complete callback to the callback queue
      * Move the queue to the last contiguous callback to commit offset
      * @param cb new callback completed
-     * @return callback of highest watermark needed to be committed
+     * @return list of callbacks to be committed
      */
-    TaskCallbackImpl update(TaskCallbackImpl cb) {
+    List<TaskCallbackImpl> update(TaskCallbackImpl cb) {
+
       synchronized (lock) {
         callbacks.add(cb);
-
         TaskCallbackImpl callback = null;
-        TaskCallbackImpl callbackToCommit = null;
+        List<TaskCallbackImpl> committableCallbacks = new ArrayList<>();
```

Comment:
Actually we don't need to have a separate commitableCallbacks. We need to update for all the callbacks below the low watermark. So this should just be callbacksToUpdate or something